fv/wireguard_test.go

```diff
@@ -0,0 +1,455 @@
+
+			for _, felix := range felixes {
+				Eventually(func() string {
+					// No tunnel implies un-encrypted communication.
+					out, _ := felix.ExecOutput("wg")
+					return out
+				}, "10s", "100ms").Should(BeEmpty())
+			}
+		})
+	})
+})
+
+var _ = infrastructure.DatastoreDescribe("WireGuard-Unsupported", []apiconfig.DatastoreType{apiconfig.EtcdV3, apiconfig.Kubernetes}, func(getInfra infrastructure.InfraFactory) {
+	var (
+		infra   infrastructure.DatastoreInfra
+		felixes []*infrastructure.Felix
+	)
+
+	BeforeEach(func() {
+		// Setup a single node cluster.
+		const nodeCount = 1
+
+		infra = getInfra()
+		felixes, _ = infrastructure.StartNNodeTopology(nodeCount, wireguardTopologyOptions(), infra)
+
+		// Install a default profile that allows all ingress and egress, in the absence of any Policy.
+		infra.AddDefaultAllow()
+
+		felixes[0].TriggerDelayedStart()
+	})
+
+	AfterEach(func() {
+		if CurrentGinkgoTestDescription().Failed {
+			felixes[0].Exec("ip", "link")
+			felixes[0].Exec("wg")
+		}
+
+		felixes[0].Stop()
+
+		if CurrentGinkgoTestDescription().Failed {
+			infra.DumpErrorData()
+		}
+		infra.Stop()
+	})
+
+	It("no Wireguard device exists", func() {
+		Eventually(func() string {
+			out, _ := felixes[0].ExecOutput("ip", "link", "show", wireguardInterfaceNameDefault)
+			return out
+		}, "10s", "100ms").Should(BeEmpty())
+	})
+
+	It("no wg tool info exists", func() {
+		Eventually(func() string {
+			out, _ := felixes[0].ExecOutput("wg")
+			return out
+		}, "10s", "100ms").Should(BeEmpty())
+	})
+})
+
+func wireguardTopologyOptions() infrastructure.TopologyOptions {
+	topologyOptions := infrastructure.DefaultTopologyOptions()
+
+	// For attaching tcpdump and waiting on calico-node to be ready.
+	topologyOptions.DelayFelixStart = true
+	// WireGuard doesn't support IPV6, disable it.
+	topologyOptions.EnableIPv6 = false
+	// Assigning workload IPs using IPAM api.
+	topologyOptions.IPIPRoutesEnabled = false
+	// Enable WireGuard.
+	felixConfig := api.NewFelixConfiguration()
+	felixConfig.SetName("default")
+	enabled := true
+	felixConfig.Spec.WireguardEnabled = &enabled
+	topologyOptions.InitialFelixConfiguration = felixConfig
+
+	// Debugging.
+	//topologyOptions.ExtraEnvVars["FELIX_DebugUseShortPollIntervals"] = "true"
+	//topologyOptions.FelixLogSeverity = "debug"
+
+	return topologyOptions
+}
+
+func enableDisableWireguard(client clientv3.Interface, enabled bool) {
+	ctx, cancel := context.WithTimeout(context.Background(), 10*time.Second)
+	defer cancel()
+	felixConfig, err := client.FelixConfigurations().Get(ctx, "default", options.GetOptions{})
+	Expect(err).NotTo(HaveOccurred())
+	felixConfig.Spec.WireguardEnabled = &enabled
+	felixConfig, err = client.FelixConfigurations().Update(ctx, felixConfig, options.SetOptions{})
+	Expect(err).NotTo(HaveOccurred())
+}
+
+func wireguardRoutingRuleRouteEntryExists(felix *infrastructure.Felix, exp *regexp.Regexp, entry string) bool {
+	// Get WireGuard route-table index.
```

Comment:
This is a very odd interface. You have a regex hardcoded into the method (which is fine for the test code), but then you pass in another regex to extract the priority.

What is exp? Why does it need to be specified?

Also, there are now potentially multiple rules that will be installed. This might be more useful if, given a "priority" and "table", it returns a list of from CIDRs.... or something like that. Something that at least allows you to check multiple rules. You can use it to check if rules have been deleted by checking for no CIDRs. That kind of thing.

Reply:
Also, the rules and routes have multiple entries and contain different data.  Wouldn't it be better to have separate validation functions for each rather than trying to combine them?

Reply:
Result from routing-rule check (i.e. table-id) feeds into the route-table entry check. For reliably checking route-entry I am using rule to feed the tableId. To avoid repeating the tableId extraction, I combined the tests in 1.

Reply:
But we know that table ID upfront don't we?  It's always the first ID to be selected from the configured range - so we don't need to extract it to then extract the rules..... or am I missing something here?

Reply:
table-id allocation is a separate code not related to WireGuard. You are right that we will see its value to be 1, at this time. But if a new table-id allocation is ever added in our code-path or if the range changes from 1-250 to some other values, this test will fail for no direct reason. I am trying to avoid that assumption by extracting the actual table-id.

I think FVs are pretty descriptive in pointing to the exact line of failure in a spec. So between the rule and route-entry failure, it should be easy to point to right failure.

Reply:
I'm fine with this for now - but we should update to check the table ID.

I explicitly agreed with Neil that Wireguard would "get in first" with the table index allocation. It is important for the table index to be deterministic to claim that we won't drop traffic unnecessarily.  Having an FV that checked that and stated it explicitly in the event of a failure would probably be a good thing.

Reply:
@realgaurav - same here. Resolved?